mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp

```diff
@@ -444,7 +444,15 @@ scf::ForOp LoopPipelinerInternal::createKernelLoop(
         loc, rewriter.getIntegerAttr(t, maxStage));
     Value maxStageByStep =
         rewriter.create<arith::MulIOp>(loc, step, maxStageValue);
-    newUb = rewriter.create<arith::SubIOp>(loc, ub, maxStageByStep);
+    Value hasAtLeastOneIteration = rewriter.create<arith::CmpIOp>(
+        loc, arith::CmpIPredicate::slt, maxStageByStep, ub);
+    Value possibleNewUB =
+        rewriter.create<arith::SubIOp>(loc, ub, maxStageByStep);
+    // In case of `index` or `unsigned` type, we need to make sure that the
+    // subtraction does not result in a negative value, instead we use lb
+    // to avoid entering the kernel loop.
+    newUb = rewriter.create<arith::SelectOp>(
+        loc, hasAtLeastOneIteration, possibleNewUB, forOp.getLowerBound());
```

Comment:
@ThomasRaoux , actualy what we want to make sure is that maxStageByStep < ub, because if those value are from `unsigned` or `index` may lead to wrong answer.
For exmaple, if:
ub = 100
maxStageByStep = 150
Then:
100 - 150 = (-50) 
But for index/unsigned this value will be intreprated to **big possitive number** as negative doesn't exist in unsigned.
This leads to error that we will actually enter the kernel loop although `ub < maxStageByStep`.
So in that case, to avoid using `ub-maxStageByStep < lb` which in the case mentioned above may be falsely true, I added a check and use `lb` just to avoid entering the kernel.
The rest of the code using `ub-maxStageByStep` is protected by `pred` condition and usually wrapped with `if` on the `pred` so I think we are o.k. there, the only place where it didn't work for me was on the `for` kernel mentioned above.

Reply:
scf.for comparison is always signed as far as I know. In what case would we have an unsigned comparison of lb and ub?

Reply:
Here is how the scf.for gets lowered:
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/SCFToControlFlow/SCFToControlFlow.cpp#L371

Reply:
The induction variable/lb/ub of `scf.for` are from `AnySignlessIntegerOrIndex` type:
https://github.com/llvm/llvm-project/blob/7e85cb8a8a9de57ed10635b843662148a87b17e5/mlir/include/mlir/Dialect/SCF/IR/SCFOps.td#L242

So it also affects the calcualtion of `ub-maxStageByStep` is `AnySignlessIntegerOrIndex`.
Index type is defined by:
https://github.com/llvm/llvm-project/blob/7e85cb8a8a9de57ed10635b843662148a87b17e5/mlir/include/mlir/IR/BuiltinTypes.td#L451
With the notion of:
`**Rationale:** integers of platform-specific bit widths are practical to
    express sizes, dimensionalities and subscripts.`

So we consider `index` as `unsigned int` in our platform.

Reply:
I don't think this is correct, signless means the type doesn't contain information on the whether it is signed or unsigned and the information is carried by the op. In the case of scf.for the comparison is signed in scf.for semantic.

Reply:
`scf.for` definition in the td file requires the `step` to be positive, but doesn't mentioned signed/unsigned. I agree that the pass you mentioned above lowers to signed, but from EmitC lowering I found lowering to `size_t` (which is similar unsigned int):
https://github.com/llvm/llvm-project/blob/abac5be673a2053cceab8ce25009722e45021b9f/mlir/lib/Target/Cpp/TranslateToCpp.cpp#L1667

@jpienaar @marbre @ThomasRaoux Personally I can agree with both signed/unsigned but I would like us to be aligned and define it in the `scf.for` td definition to make it clear. With the current status my platform lowers `scf.for` lb/ub/step into `unsigned int`

Reply:
I agree we should clarify the definition of scf.for. The comparison needs to be signed or unsigned it cannot be either. 
I would suggest we clarify the spec to clearly mention that each iteration does the signed comparison `ind < ub`

Reply:
@ThomasRaoux @joker-eph 
After some discussion iwth my team it seems more reasonable to make the lower/upper bounds as `signed integer` when lowered to HW.
Here is a followed up PR: https://github.com/llvm/llvm-project/pull/117534